FreeRTOS/FreeRTOS-Plus-TCP
# Add port for ATSAME5x (Atmel SAM E5x)

Atmel SAM E5x Network Interface

Description
-----------
Add a port (NetworkInterface.c) for ATSAME5x devices (Atmel SAM E5x). The port is based on Atmel's ASF4 GMAC driver, which remains unmodified. This port is NOT zero-copy and is generally meant to prioritize small size over performance.

Currently, configuration is done through Atmel START in Microchip Studio 7.0 and is stored in auto-generated file "hpl_gmac_config.h". 

Test Steps
-----------
See test project: https://github.com/ma-lalonde/ATSAME54N20A_FreeRTOS_Plus_TCP
This is developed for a custom ATSAME54N20A board, where there is a blinking LED for quick debugging (if the application freezes / crashes, the LED stops blinking).

Completed tests include DHCP, ICMP Echo (ping) send and receive (validated with WireShark, including CRC check), LLMNR (using the MCU's unique ID as network name), and iPerf3 testing using @htibosch's iperf3 server. Speed result is 14-15 Mbps with both BufferAllocation_1.c and BufferAllocation_2.c with CPU clock at 96 MHz, which makes me believe the true bottleneck is not being zero-copy. 

Related Issue
-----------
I needed the driver, so I made it, with some help: https://forums.freertos.org/t/re-freertos-tcp-on-sam-e5x


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
